### ISSUE TITLE
Performance degradation for Z3

### ISSUE THREAD

**Issue by rafaelsamenezes** [esbmc/esbmc#752]
The above is for the large array https://github.com/esbmc/esbmc/blob/c7191cda390244cabca0285d0c41a7ad2ffa673e/regression/esbmc/github_732-2-2/main.c

Other solvers do the `--smt-formula-only` much quicker and with less memory:
- Boolector 7.2s 144M
- CVC4 0.77s 152M
- Yices 7.1s 238M
- Bitwuzla 8.9s 146M
- Mathsat 1.4s 175M
- well... and Z3 116.7s 640M

_Originally posted by @fbrausse in https://github.com/esbmc/esbmc/issues/749#issuecomment-1156257025_

**Comment by lucasccordeiro** [comment]
@fbrausse: can I ask you whether you could look at this issue? This is a low-priority issue.

**Comment by fbrausse** [comment]
`perf report -g` output:
```
# Children      Self  Command  Shared Object                       Symbol                                                                                                                                                                                                       >
# ........  ........  .......  ..................................  .............................................................................................................................................................................................................>
#
    65.70%    65.65%  esbmc    libz3.so.4.8.17.0                   [.] ast_manager::~ast_manager
            |          
             --18.10%--ast_manager::~ast_manager

    19.85%    19.84%  esbmc    libz3.so.4.8.17.0                   [.] ast_mark::is_marked
            |          
             --4.55%--ast_mark::is_marked

    11.91%    11.91%  esbmc    libz3.so.4.8.17.0                   [.] ast_mark::mark
     5.11%     0.00%  esbmc    [unknown]                           [.] 0000000000000000
            |
            ---0
               |          
               |--2.64%--ast_mark::is_marked
               |          
                --2.05%--ast_manager::~ast_manager

     4.29%     0.00%  esbmc    [unknown]                           [.] 0x0000000000000001
            |
            ---0x1
               |          
               |--2.38%--ast_mark::is_marked
               |          
                --1.90%--ast_manager::~ast_manager

     0.43%     0.41%  esbmc    libz3.so.4.8.17.0                   [.] ast_manager::register_node_core
     0.35%     0.00%  esbmc    libc.so.6                           [.] __libc_start_call_main
     0.35%     0.00%  esbmc    esbmc                               [.] main
     0.35%     0.00%  esbmc    esbmc                               [.] parseoptions_baset::main
     0.35%     0.00%  esbmc    esbmc                               [.] esbmc_parseoptionst::doit
     0.34%     0.00%  esbmc    esbmc                               [.] smt_convt::convert_ast
```

**Comment by fbrausse** [comment]
This is the formula as generated by --smt-formula-too for Z3: [gh-732-2-2.txt](https://github.com/esbmc/esbmc/files/8942240/gh-732-2-2.txt)

Interesting that the `perf` output for `z3 -v:15 gh-732-2-2.smt2` looks different: `ast_manager::~ast_manager` is not even part of it. <details>

```
# Children      Self  Command  Shared Object         Symbol                                                                                                                                                                                                                     >
# ........  ........  .......  ....................  ...........................................................................................................................................................................................................................>
#
    26.44%     0.00%  z3       [unknown]             [.] 0xffffffff00000001
            |
            ---0xffffffff00000001
               |          
                --26.16%--(anonymous namespace)::th_rewriter_cfg::reduce_app
                          |          
                          |--24.54%--array_rewriter::mk_select_core
                          |          |          
                          |          |--16.56%--ast_manager::register_node_core
                          |          |          
                          |          |--4.37%--small_object_allocator::allocate
                          |          |          
                          |          |--0.85%--ast_manager::mk_app_core
                          |          |          
                          |          |--0.67%--ast_array_hash<expr>
                          |          |          
                          |           --0.51%--get_node_hash
                          |          
                           --0.72%--ast_manager::are_distinct

    26.27%     0.52%  z3       z3                    [.] (anonymous namespace)::th_rewriter_cfg::reduce_app
            |          
             --25.75%--(anonymous namespace)::th_rewriter_cfg::reduce_app
                       |          
                       |--24.54%--array_rewriter::mk_select_core
                       |          |          
                       |          |--16.56%--ast_manager::register_node_core
                       |          |          
                       |          |--4.37%--small_object_allocator::allocate
                       |          |          
                       |          |--0.85%--ast_manager::mk_app_core
                       |          |          
                       |          |--0.67%--ast_array_hash<expr>
                       |          |          
                       |           --0.51%--get_node_hash
                       |          
                        --0.72%--ast_manager::are_distinct

    24.59%     0.87%  z3       z3                    [.] array_rewriter::mk_select_core
            |          
             --23.71%--array_rewriter::mk_select_core
                       |          
                       |--16.56%--ast_manager::register_node_core
                       |          
                       |--4.37%--small_object_allocator::allocate
                       |          
                       |--0.85%--ast_manager::mk_app_core
                       |          
                       |--0.67%--ast_array_hash<expr>
                       |          
                        --0.51%--get_node_hash

    21.41%     0.00%  z3       [unknown]             [.] 0x00005629bc358868
            |
            ---0x5629bc358868
               |          
               |--11.23%--act_cache::find
               |          
               |--8.05%--bv_decl_plugin::is_unique_value
               |          
                --0.81%--rewriter_tpl<(anonymous namespace)::th_rewriter_cfg>::visit<false>

    18.50%    18.48%  z3       z3                    [.] act_cache::del_unused
    17.75%    17.73%  z3       z3                    [.] ast_manager::register_node_core
            |          
             --1.13%--0
                       ast_manager::register_node_core

    15.65%     0.00%  z3       [unknown]             [k] 0000000000000000
            |
            ---0
               |          
               |--5.60%--0x8000000410000
               |          |          
               |          |--3.90%--ast_table::push_erase
               |          |          
               |           --1.38%--ast_manager::delete_node
               |          
               |--4.26%--simplify_tactic::imp::operator()
               |          |          
               |           --3.66%--rewriter_tpl<(anonymous namespace)::th_rewriter_cfg>::operator()
               |          
               |--1.13%--ast_manager::register_node_core
               |          
               |--0.78%--bv_decl_plugin::get_offset_term
               |          
               |--0.71%--array_decl_plugin::mk_select
               |          
                --0.51%--mpz_manager<true>::del

    13.75%     0.00%  z3       [unknown]             [.] 0x00005629be7d4fc8
            |
            ---0x5629be7d4fc8
               |          
               |--10.11%--act_cache::del_unused
               |          
                --3.51%--act_cache::insert

    12.42%     0.00%  z3       [unknown]             [.] 0xec8348fd89485355
            |
            ---0xec8348fd89485355
               th_rewriter::imp::~imp
               0x5629bc358868
               |          
               |--11.23%--act_cache::find
               |          
                --0.81%--rewriter_tpl<(anonymous namespace)::th_rewriter_cfg>::visit<false>

    12.42%     0.00%  z3       z3                    [.] th_rewriter::imp::~imp
            |
            ---th_rewriter::imp::~imp
               0x5629bc358868
               |          
               |--11.23%--act_cache::find
               |          
                --0.81%--rewriter_tpl<(anonymous namespace)::th_rewriter_cfg>::visit<false>

    11.48%    11.48%  z3       z3                    [.] act_cache::find
            |          
             --11.23%--0xec8348fd89485355
                       th_rewriter::imp::~imp
                       0x5629bc358868
                       act_cache::find

    10.96%     0.00%  z3       [unknown]             [.] 0x00005629be0560a8
            |
            ---0x5629be0560a8
               |          
               |--7.53%--act_cache::del_unused
               |          
                --3.28%--act_cache::insert

     8.97%     0.00%  z3       z3                    [.] decl_plugin::inherit
            |
            ---decl_plugin::inherit
               0x5629bc358868
               |          
                --8.05%--bv_decl_plugin::is_unique_value

     8.70%     0.00%  z3       [unknown]             [k] 0xfd894853555441f6
            |
            ---0xfd894853555441f6
               decl_plugin::inherit
               0x5629bc358868
               |          
                --8.05%--bv_decl_plugin::is_unique_value

     8.05%     8.04%  z3       z3                    [.] bv_decl_plugin::is_unique_value
            |          
             --8.04%--0xfd894853555441f6
                       decl_plugin::inherit
                       0x5629bc358868
                       bv_decl_plugin::is_unique_value

     7.27%     7.26%  z3       z3                    [.] act_cache::insert
     5.60%     0.00%  z3       [unknown]             [.] 0x0008000000410000
            |
            ---0x8000000410000
               |          
               |--3.90%--ast_table::push_erase
               |          
                --1.38%--ast_manager::delete_node

     5.26%     5.25%  z3       z3                    [.] rewriter_tpl<(anonymous namespace)::th_rewriter_cfg>::operator()
     4.54%     4.53%  z3       z3                    [.] small_object_allocator::allocate
     4.26%     0.00%  z3       z3                    [.] simplify_tactic::imp::operator()
            |
            ---simplify_tactic::imp::operator()
               |          
                --3.66%--rewriter_tpl<(anonymous namespace)::th_rewriter_cfg>::operator()

     3.90%     3.90%  z3       z3                    [.] ast_table::push_erase
            |          
             --3.90%--0
                       0x8000000410000
                       ast_table::push_erase

     2.13%     0.26%  z3       z3                    [.] ast_manager::mk_app
            |          
             --1.87%--ast_manager::mk_app
                       |          
                       |--1.11%--decl_plugin::mk_func_decl
                       |          
                        --0.51%--array_decl_plugin::mk_select
```
</details>

**Comment by fbrausse** [comment]
I think this solves the problem:
```diff
diff --git a/src/solvers/z3/z3_conv.cpp b/src/solvers/z3/z3_conv.cpp
index 9af222da29..ac5c4f1835 100644
--- a/src/solvers/z3/z3_conv.cpp
+++ b/src/solvers/z3/z3_conv.cpp
@@ -767,14 +767,14 @@ smt_astt z3_convt::mk_concat(smt_astt a, smt_astt b)
   smt_sortt s =
     mk_bv_sort(a->sort->get_data_width() + b->sort->get_data_width());
 
-  return new_ast(
-    z3::to_expr(
-      z3_ctx,
-      Z3_mk_concat(
-        z3_ctx,
-        to_solver_smt_ast<z3_smt_ast>(a)->a,
-        to_solver_smt_ast<z3_smt_ast>(b)->a)),
-    s);
+  Z3_ast z3a = Z3_mk_concat(
+    z3_ctx,
+    to_solver_smt_ast<z3_smt_ast>(a)->a,
+    to_solver_smt_ast<z3_smt_ast>(b)->a);
+
+  smt_astt r = new_ast(z3::to_expr(z3_ctx, z3a), s);
+  Z3_dec_ref(z3_ctx, z3a);
+  return r;
 }
 
 smt_astt z3_convt::mk_ite(smt_astt cond, smt_astt t, smt_astt f)
```
If so, of course it should be applied throughout z3_conv.

The core of it: `z3::to_expr()` increases the reference count on the `Z3_ast` expression. Therefore, after wrapping it into `z3::to_expr` we don't need the reference anymore and should release it. By not doing that at present, we cause a memory leak that gets cleaned up (in a probably improvable way since it takes so long) by the dtor `ast_manager::~ast_manager` at the very end.

**Comment by fbrausse** [comment]
> I think this solves the problem:

Actually not, sorry for the noise. Appearently they use a floating reference count initially on creation of ASTs, similar to GTK+'s [GObject](https://docs.gtk.org/gobject/concepts.html#object-memory-management).

Investigating further...

**Comment by fbrausse** [comment]
It has something to do with memory management of AST nodes. ESBMC's or Z3's. In https://github.com/fbrausse/esbmc/compare/fb/fix-752 the smt_asts are deleted before destructing the z3::context, which circumvents this problem. The root cause is still unknown. I'll open a PR after #749 is merged since this change reverts marking the new test case THOROUGH.

**Comment by lucasccordeiro** [comment]
Thanks, @fbrausse. I'll merge this PR https://github.com/esbmc/esbmc/pull/749 to progress further with the performance fix, which is crucial for our industrial collaborator.

**Comment by lucasccordeiro** [comment]
@fbrausse: I have just merged https://github.com/esbmc/esbmc/pull/749. Could you please submit the PR to fix the performance degradation with Z3?

**Comment by fbrausse** [comment]
Though the solving times get much more normal, [appearently](https://github.com/fbrausse/esbmc/runs/6982525300?check_suite_focus=true) the memory problem persists. :/

**Comment by fbrausse** [comment]
Using the --smt-formula-too output, it is possible to reproduce the huge memory requirements (though not the timing ones):
```
155.26user 1.59system 2:36.95elapsed 99%CPU (0avgtext+0avgdata 9148928maxresident)k
24464inputs+0outputs (138major+2298988minor)pagefaults 0swaps
```
Watching the process it seems that nearly all the time it uses ~100M and only at the very end jumps to 9G.

I'll file a bug with Z3.

**Comment by fbrausse** [comment]
With latest master of ESBMC and Z3 this issue does not exist anymore. What is the policy on updating dependencies like Z3? 

Btw. there are also newer (minor) releases of Boolector: [3.2.1 and 3.2.2](https://github.com/Boolector/boolector/releases).

**Comment by lucasccordeiro** [comment]
@rafaelsamenezes: could you check these newer versions of Boolector? We should keep our regression over the SV-COMP benchmarks. In the past, we had situations where the newer Boolector versions suffered significant performance degradation in SV-COMP.

**Comment by rafaelsamenezes** [comment]
Ok, lots of requests here:

> With latest master of ESBMC and Z3 this issue does not exist anymore. What is the policy on updating dependencies like Z3?

Since we do not build Z3 at the CI, it will not be trivial as cvc4 and boolector. I will be adding it though. https://github.com/esbmc/esbmc/issues/761

> Btw. there are also newer (minor) releases of Boolector: [3.2.1 and 3.2.2](https://github.com/Boolector/boolector/releases).

> could you check these newer versions of Boolector? We should keep our regression over the SV-COMP benchmarks. In the past, we had situations where the newer Boolector versions suffered significant performance degradation in SV-COMP.

This is more straightforward to do, just change the branch here: https://github.com/esbmc/esbmc/blob/master/.github/workflows/build.yml#L65

We can then open a PR and run sv-comp directly.

**Comment by rafaelsamenezes** [comment]
These are the current failing tests for Z3. We still have some performance issues.

These tests were executed at:
- Ubuntu 20.04
- Z3 4.8.9 (still has the memory bug)
- I've compiled ESBMC  only with Z3.  

```
The following tests FAILED:
         24 - regression/esbmc-unix/00_account_01 (Failed)
         25 - regression/esbmc-unix/00_account_02 (Failed)
         76 - regression/esbmc-unix/01_malloc_08 (Timeout)
        327 - regression/esbmc/00_big_endian_01 (Failed)
        337 - regression/esbmc/00_little_endian_04 (Failed)
        340 - regression/esbmc/00_memcpy_04 (Failed)
        402 - regression/esbmc/12_lms_new (Failed)
        458 - regression/esbmc/builtin_init_var_3_false (Failed)
        462 - regression/esbmc/builtin_init_var_5_false (Failed)
        549 - regression/esbmc/github_302 (Timeout)
        551 - regression/esbmc/github_306 (Timeout)
        626 - regression/esbmc/github_629 (Failed)
        679 - regression/esbmc/math_cos01 (Failed)
        684 - regression/esbmc/math_log02 (Failed)
        712 - regression/esbmc/overflow_10_fir (Failed)
        717 - regression/esbmc/overflow_16_fir_new (Failed)
        758 - regression/esbmc/to_union_1 (Failed)
        763 - regression/esbmc/to_union_github_314_float (Failed)
        764 - regression/esbmc/union_assignment (Failed)
        1161 - regression/floats/Float-data-dependent-rounding (Failed)
        1165 - regression/floats/Float-flags-no-simp1 (Failed)
        1171 - regression/floats/Float-no-simp4 (Failed)
        1172 - regression/floats/Float-no-simp4_1 (Failed)
        1173 - regression/floats/Float-no-simp5 (Failed)
        1176 - regression/floats/Float-no-simp8 (Failed)
        1177 - regression/floats/Float-no-simp9 (Failed)
        1182 - regression/floats/Float-zero-sum1 (Failed)
        1190 - regression/floats/Float19 (Failed)
        1194 - regression/floats/Float22 (Failed)
        1197 - regression/floats/Float4 (Failed)
        1205 - regression/floats/Float_lib2 (Failed)
        1216 - regression/floats-regression/ceil (Failed)
        1217 - regression/floats-regression/ceil_nondet (Failed)
        1218 - regression/floats-regression/copysign (Failed)
        1224 - regression/floats-regression/fdim (Failed)
        1225 - regression/floats-regression/floor (Failed)
        1226 - regression/floats-regression/floor_nondet (Failed)
        1228 - regression/floats-regression/fmax (Failed)
        1229 - regression/floats-regression/fmin (Failed)
        1230 - regression/floats-regression/fmod (Failed)
        1231 - regression/floats-regression/fmod2 (Failed)
        1232 - regression/floats-regression/fmod3 (Failed)
        1239 - regression/floats-regression/lrint (Failed)
        1240 - regression/floats-regression/modf (Failed)
        1244 - regression/floats-regression/nextafter (Failed)
        1245 - regression/floats-regression/nextafter-spec (Failed)
        1246 - regression/floats-regression/nextafterf-spec (Failed)
        1247 - regression/floats-regression/remainder (Failed)
        1249 - regression/floats-regression/rint (Failed)
        1250 - regression/floats-regression/rint2 (Failed)
        1251 - regression/floats-regression/round (Failed)
        1252 - regression/floats-regression/round_nondet (Failed)
        1253 - regression/floats-regression/rounding_functions (Failed)
        1255 - regression/floats-regression/trunc (Failed)
        1256 - regression/floats-regression/trunc_nondet (Failed)
        1257 - regression/floats-regression/trunc_nondet_2 (Failed)
        1258 - regression/floats-regression/union_inc (Failed)
        1259 - regression/floats-regression/union_inc2 (Failed)
        1296 - regression/k-induction/problem01_20_unsafe (Timeout)
        1786 - regression/k-induction-parallel/problem01_20_unsafe (Timeout)
        1837 - regression/nonz3/02_eureka02 (Failed)
        1838 - regression/nonz3/02_eureka21 (Failed)
        1846 - regression/nonz3/overflow_12_adpcm (Failed)
        1874 - regression/incremental-smt/incremental-26 (Timeout)
```

**Comment by fbrausse** [comment]
> These are the current failing tests for Z3. We still have some performance issues

A little more info, please. What build, what tests were run, what platform, how can I reproduce any of this? For instance (picking at random):
```
$ env ESBMC_OPTS=--z3 ctest -R regression/floats-regression/fmax
Test project /home/kane/scorch/esbmc/debug
    Start 1230: regression/floats-regression/fmax
1/1 Test #1230: regression/floats-regression/fmax ...   Passed    0.74 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.86 sec
```

**Comment by rafaelsamenezes** [comment]
> A little more info, please.

I've added to the original comment

**Comment by fbrausse** [comment]
> - I've compiled ESBMC only with Z3.

@rafaelsamenezes Your list is not usable at the moment, for instance it contains
```
        1851 - regression/incremental-smt/incremental-03 (Failed)
```
which has `--bitwuzla` explicitely https://github.com/esbmc/esbmc/blob/e2113d2a54acc456ebc801ede3c9c05341ae33f1/regression/incremental-smt/incremental-03/test.desc#L1-L4 and ESBMC neither likes a) specifying multiple solvers nor b) running with a non-built-in solver.

> - Z3 4.8.9 (still has the memory bug)

These are not related to this issue. This issue was about a specific performance bug in Z3, which is fixed in ESBMC and Z3 master.

**Comment by rafaelsamenezes** [comment]
> These are not related to this issue.

Ow, I will remove those.

> This issue was about a specific performance bug in Z3, which is fixed in ESBMC and Z3 master.

The issue is still happening. The issue is that we cannot pinpoint any test case.

**Comment by fbrausse** [comment]
Where do you see this issue happening? For me it is fixed:
```
$ /usr/bin/time src/esbmc/esbmc --z3 ../regression/esbmc/github_732-2-2/main.c 
ESBMC version 6.10.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmc libc
[...]
Solving with solver Z3 v4.8.18
Encoding to solver time: 0.491s
[...]
VERIFICATION FAILED
Command exited with non-zero status 1
3.64user 0.57system 0:04.22elapsed 99%CPU (0avgtext+0avgdata 1839688maxresident)k
0inputs+0outputs (0major+670908minor)pagefaults 0swaps
```

**Comment by rafaelsamenezes** [comment]
> Where do you see this issue happening? For me it is fixed:

That's the problem, I am still trying to get a test case that showcase this (this is reported by our stakeholders). The only thing I know is that reverting https://github.com/esbmc/esbmc/pull/615 fixes the performance for them. 

Should I close this and reopen for this specific reason?

**Comment by fbrausse** [comment]
> That's the problem, I am still trying to get a test case that showcase this (this is reported by our stakeholders). The only thing I know is that reverting https://github.com/esbmc/esbmc/pull/615 fixes the performance for them.
>
> Should I close this and reopen for this specific reason?

Depends. Is there a connection to the issue about memory/performance with `regression/esbmc/github_732-2-2` that we dealt with initially? It could be, since #615 did lead to encoding unions as long BVs, which also (somehow) were involved in this issue. Though I doubt it is really the same problem: the bug here was caused by excessively long bit-vectors due to arrays in unions being flattened on the SMT level. The size of arrays is given logarithmically, which caused our stack-overflow and Z3's memory problem for large-ish arrays. Since the cause for this one here has been found and fixed (actually, we didn't enable the `THOROUGH` test for Z3, yet, we should), I expected that we can close this after v.4.8.18 of Z3 is released and we update our build instructions...

It was clear that #615 itself changes a lot of things in ESBMC. Some could affect Z3. I didn't check back then. Maybe it is worth opening an issue about it. We could, for instance, try running the SV-COMP benchmarks on a default-Z3 build of ESBMC and see whether that affects timings...

**Comment by fbrausse** [comment]
@rafaelsamenezes On the [fb/tst-z3](https://github.com/esbmc/esbmc/tree/fb/tst-z3) branch I've allowed via the benchexec action to pass `--default-solver z3` as cmdline options, but I don't have a Ubuntu-20.04 build of Z3 master to compile ESBMC against in this workflow. If you had such a Z3 build, we could start checking SV-COMP for regressions with Z3.

**Comment by rafaelsamenezes** [comment]
> On the [fb/tst-z3](https://github.com/esbmc/esbmc/tree/fb/tst-z3) branch I've allowed via the benchexec action to pass --default-solver z3 as cmdline options, but I don't have a Ubuntu-20.04

Selecting which solver to use looks like a good contribution for the benchexec action. 

>  I don't have a Ubuntu-20.04 build of Z3 master to compile ESBMC against in this workflow. If you had such a Z3 build, we could start checking SV-COMP for regressions with Z3.

I also do not have one, I was waiting for the next Z3 official release. However, I can prepare one this afternoon.

**Comment by fbrausse** [comment]
> Selecting which solver to use looks like a good contribution for the benchexec action.

Thanks! It allows arbitrary cmdline options. I'll open a PR.

>  I also do not have one, I was waiting for the next Z3 official release.

Same here :)

> However, I can prepare one this afternoon.

That would be great, many thanks! I don't have an Ubuntu 20.04 anywhere close by right now.

**Comment by fbrausse** [comment]
It should now be possible to investigate this further via the benchmarks with `--z3`, as well as the regression suite with `ESBMC_OPTS` set to `--default-solver z3`. The latter will use Z3 when no solver is explicitely specified in the regression test. This won't test all of them, still a majority though, which I believe is intended.

**Comment by fbrausse** [comment]
Problematic ones from the regression suite locally for current master (3eb0a4f154d49f9f67b825d950aed55244ff7c19) with 15m timeout:
```
99% tests passed, 11 tests failed out of 1937

Total Test time (real) = 1569.58 sec

The following tests FAILED:
	458 - regression/esbmc/builtin_init_var_3_false (Failed)
	462 - regression/esbmc/builtin_init_var_5_false (Failed)
	549 - regression/esbmc/github_302 (Timeout)
	551 - regression/esbmc/github_306 (Timeout)
	628 - regression/esbmc/github_629 (Timeout)
	1749 - regression/csmith/csmith01 (Timeout)
	1776 - regression/k-induction-parallel/invert_string (Failed)
	1837 - regression/nonz3/02_eureka02 (Failed)
	1838 - regression/nonz3/02_eureka21 (Failed)
	1846 - regression/nonz3/overflow_12_adpcm (Failed)
	1888 - regression/incremental-smt/incremental-26 (Timeout)
```
Edit:
- Linux (Gentoo)
- Z3-4.8.18.0, 41deed59a3201375c5c31513c06a4020f594e2a5
- usual other solvers are built in: `Available solvers: smtlib z3 boolector cvc4 mathsat yices bitwuzla`

**Comment by fbrausse** [comment]
All of the failing ones except `regression/nonz3/overflow_12_adpcm` crash with some assertion error about widths in smt_conv. The one that doesn't wrongly gives
```
State 134 file main.c line 209 function decode thread 0
----------------------------------------------------
Violated property:
  file main.c line 209 function decode
  arithmetic overflow on mul
  !overflow("*", (signed long int)dec_detl, (signed long int)(qq6_code6_table[(signed long int)il]))


VERIFICATION FAILED
```

Is one of the timeouts maybe responsible for whereever you see this issue here happening?

**Comment by fbrausse** [comment]
The performance fix for Z3 has been been part of releases since v4.9 on 06 July. IMO this was the last component remaining to fix this issue. @rafaelsamenezes Could I ask you to open a new issue once you find any concrete example where ESBMC/Z3 has definable problems? Then we could finally close this issue, which has been about a pretty clearly defined problem.

**Comment by rafaelsamenezes** [comment]
I agree, we may close this issue.
